yamux/src/connection/stream.rs

```diff
@@ -360,54 +362,65 @@ impl AsyncWrite for Stream {
         ready!(self
             .sender
             .poll_ready(cx)
-            .map_err(|_| self.write_zero_err())?);
-        let body = {
-            let mut shared = self.shared();
-            if !shared.state().can_write() {
-                log::debug!("{}/{}: can no longer write", self.conn, self.id);
-                return Poll::Ready(Err(self.write_zero_err()));
+            .map_err(|_| Stream::write_zero_err(self.conn, self.id))?);
+
+        let stream = self.as_mut().get_mut();
+        let result = stream.shared.with_mut(|inner| {
+            if !inner.state.can_write() {
+                log::debug!("{}/{}: can no longer write", stream.conn, stream.id);
+                return Err(Stream::write_zero_err(stream.conn, stream.id));
             }
-            if shared.send_window() == 0 {
-                log::trace!("{}/{}: no more credit left", self.conn, self.id);
-                shared.writer = Some(cx.waker().clone());
-                return Poll::Pending;
+
+            let window = inner.send_window();
+            if window == 0 {
+                log::trace!("{}/{}: no more credit left", stream.conn, stream.id);
+                inner.writer = Some(cx.waker().clone());
+                return Ok(None);
             }
-            let k = std::cmp::min(
-                shared.send_window(),
-                buf.len().try_into().unwrap_or(u32::MAX),
-            );
+
+            let k = std::cmp::min(window, buf.len().try_into().unwrap_or(u32::MAX));
             let k = std::cmp::min(
                 k,
-                self.config.split_send_size.try_into().unwrap_or(u32::MAX),
+                stream.config.split_send_size.try_into().unwrap_or(u32::MAX),
             );
-            shared.consume_send_window(k);
-            Vec::from(&buf[..k as usize])
+
+            inner.consume_send_window(k);
+            Ok(Some(Vec::from(&buf[..k as usize])))
+        });
+
+        let body = match result {
+            Err(e) => return Poll::Ready(Err(e)),
+            Ok(None) => return Poll::Pending,
+            Ok(Some(b)) => b,
         };
+
         let n = body.len();
-        let mut frame = Frame::data(self.id, body).expect("body <= u32::MAX").left();
-        self.add_flag(frame.header_mut());
-        log::trace!("{}/{}: write {} bytes", self.conn, self.id, n);
+        let mut frame = Frame::data(stream.id, body)
+            .expect("body <= u32::MAX")
+            .left();
+
+        stream.add_flag(frame.header_mut());
+
+        log::trace!("{}/{}: write {} bytes", stream.conn, stream.id, n);
 
-        // technically, the frame hasn't been sent yet on the wire but from the perspective of this data structure, we've queued the frame for sending
-        // We are tracking this information:
-        // a) to be consistent with outbound streams
-        // b) to correctly test our behaviour around timing of when ACKs are sent. See `ack_timing.rs` test.
```

Comment:
I don't think this comment should be removed?

Reply:
Done.